app/src/main/java/fr/free/nrw/commons/explore/map/ExploreMapFragment.java

```diff
@@ -354,13 +353,14 @@ private void performMapReadyActions() {
             mapView.getOverlayManager().getTilesOverlay()
                 .setColorFilter(TilesOverlay.INVERT_COLORS);
         }
-        if (!applicationKvStore.getBoolean("doNotAskForLocationPermission", false) ||
-            PermissionUtils.hasPermission(getActivity(),
-                new String[]{Manifest.permission.ACCESS_FINE_LOCATION})) {
-            checkPermissionsAndPerformAction();
-        } else {
+        if (applicationKvStore.getBoolean("doNotAskForLocationPermission", false) &&
```

Comment:
Actually, `doNotAskForLocationPermission` being set to true has been removed only in 2 instances in my PR (one in each `ExploreMapFragment` and `NearbyFragment`) because I have changed the logic of the flow. But, it is still being set to true in `ContributionsFragment`, when the user clicks no in the dialog for `showNearbyCardPermissionRationale` and this is the correct place for it to be set to true, as user doesn't want use to ask for permission.

Reply:
I agree that setting it to `true` in the showNearbyCardPermissionRationale is a right thing to do. But that can't be the only place where we should be setting this to `true` right? For instance, the user could possibly hit "Yes" for the dialog and later decide not to give permission to the app in which case `doNotAskForLocationPermission` wouldn't be set to `true`. 

Also, there are more instances of code that seem to set it to `false` now. So, I think it seems inconsistent that we're setting it to `true` in just one place, isn't it?

PS: These kind of global  states are actually causing more chaos with reasoning about the flow of code. It would be ideal if we didn't have to deal with them. That's a totally different discussion altogether, though.

Reply:
Um, I can think of only one case when `doNotAskForLocationPermission` should be set to true.

> I agree that setting it to `true` in the showNearbyCardPermissionRationale is a right thing to do. But that can't be the only place where we should be setting this to `true` right? For instance, the user could possibly hit "Yes" for the dialog and later decide not to give permission to the app in which case `doNotAskForLocationPermission` wouldn't be set to `true`.

Um, if user hits "yes" and later denies permission twice, it will be handled automatically by my code logic. So that's not a problem. Maybe understanding the flow for devs can be one. To solve that should I add a comment somewhere in the code (maybe in LocationPermissionsHelper)?

> Also, there are more instances of code that seem to set it to `false` now. So, I think it seems inconsistent that we're setting it to `true` in just one place, isn't it?
>
> PS: These kind of global states are actually causing more chaos with reasoning about the flow of code. It would be ideal if we didn't have to deal with them. That's a totally different discussion altogether, though.

I agree with that, but this global state is just for if the user clicks "No", in the dialog. We can further discuss if this is actually required or just causes and extra step and adds complexity to the flow :)

Reply:
I think there's some confusion as to the purpose of that particular "Permission request" dialog. That dialog is actually shown to ask consent to the user if it is fine to get the location and show the nearby card. The actual state of that preference is handled via the `displayNearbyCardView` key.

I think the purpose of the `doNotAskForLocationPermission` is far beyond the scope of that dialog. That dialog just turns out to be one instance in which we set it to `true`. To be honest I'm not very sure what it's purpose is. Better not fix what isn't broken. So, could you check the feasibility of restoring the excess removal of the setters of `doNotAskForLocationPermission` ? Or if it seems to cause a particular problem, could you elaborate on the same?

Reply:
> I think there's some confusion as to the purpose of that particular "Permission request" dialog. That dialog is actually shown to ask consent to the user if it is fine to get the location and show the nearby card. The actual state of that preference is handled via the `displayNearbyCardView` key.

Um, but `displayNearbyCardView` is being set to false only when onSwipe is called, not also when the user clicks "No" on the dialog. So that means if the user clicks "No" on the dialog, but still gives permission (maybe by going to the settings of the app), then also the nearby card will be shown :(

> I think the purpose of the `doNotAskForLocationPermission` is far beyond the scope of that dialog. That dialog just turns out to be one instance in which we set it to `true`. To be honest I'm not very sure what it's purpose is. Better not fix what isn't broken. So, could you check the feasibility of restoring the excess removal of the setters of `doNotAskForLocationPermission` ? Or if it seems to cause a particular problem, could you elaborate on the same?

In the main branch, `doNotAskForLocationPermission` was set in 5 cases. Once in that dialog, and twice each in Explore and Nearby map fragments. From what I understand, this key is used to handle cases when the user has denied permission twice (which means permission won't be asked for again) or has clicked No in the dialog box. I have already handled the user denying permission twice with a different logic, so now the role of this key is reduced to only when the dialog is clicked No.

I have removed the instances of it being set in cases of 
1. startMapWithCondition of `NearbyParentFragment`: because I am no longer using the method startMapWithCondition.
2. startMapWithoutPermission of` ExploreMapFragment`: because this method is called even when the user denies permission only once, which doesn't mean that we do not ask for permission since we can ask again (and thus can't set `doNotAskForLocationPermission` to true)

Reply:
Thanks for the details explanation. This is valuable information that could potentially be present in the commit message. If possible, kindly include the same in some way.

That said, I just have one concern. If we're clear that the use of the `doNotAskForLocationPermission` why are we still setting the same to `false` in some cases in the `NearbyParentFragment` and `ExploreMapFragment` fragments. It's just not adding up for me that we set it to `false` but don't set it to `true` in corresponding cases.

If there's no specific reason, I believe we could remove them too.

Reply:
> Thanks for the detailed explanation. This is valuable information that could potentially be present in the commit message. If possible, kindly include the same in some way.

I'm thinking of adding a detailed explanation of the flow in `LocationPermissionsHelper` as a comment. What would you suggest?

> That said, I just have one concern. If we're clear that the use of the `doNotAskForLocationPermission` why are we still setting the same to `false` in some cases in the `NearbyParentFragment` and `ExploreMapFragment` fragments. It's just not adding up for me that we set it to `false` but don't set it to `true` in corresponding cases.
> 
> If there's no specific reason, I believe we could remove them too.

Um, from what I understand, it's being set to false inside the `locationPermissionGranted()` of both `NearbyParentFragment` and `ExploreMapFragment` to handle the case when the user clicks "No" in the particular dialog, but later gives permission through the app's settings.

Reply:
> I'm thinking of adding a detailed explanation of the flow in LocationPermissionsHelper as a comment. What would you suggest?

Feel free to do so.

> Um, from what I understand, it's being set to false inside the locationPermissionGranted() of both NearbyParentFragment and ExploreMapFragment to handle the case when the user clicks "No" in the particular dialog, but later gives permission through the app's settings.

That makes sense. My question is if we need concern about the case where the user gives permission through app's settings, do we also not need to concern ourselves about the case where user clicks "Yes" and later revokes permission through the settings screen. In this case the `doNotAskForLocationPermission` should be set to `true` from NearbyParentFragment and ExploreMapFragment too right?

Note well that I'm only suggesting to set it to `true` in the inverse case. I don't suggest using the same in those fragments to handle the flow.

Reply:
> > I'm thinking of adding a detailed explanation of the flow in LocationPermissionsHelper as a comment. What would you suggest?
> 
> Feel free to do so.

Sorry for responding late, am caught up in my university mid semester exams. I'll add these in a couple of days.

> That makes sense. My question is if we need concern about the case where the user gives permission through app's settings, do we also not need to concern ourselves about the case where user clicks "Yes" and later revokes permission through the settings screen. In this case the `doNotAskForLocationPermission` should be set to `true` from NearbyParentFragment and ExploreMapFragment too right?

Um, figuring out when user goes to settings and revokes the permission would be difficult. Also, Android turns off permission automatically for unused apps, so in this case if we set `doNotAskForLocationPermission` to true, then we will never be able to ask for permission even when user hasn't himself denied or turned it off.

Reply:
> > I'm thinking of adding a detailed explanation of the flow in LocationPermissionsHelper as a comment. What would you suggest?
> 
> Feel free to do so.

Sorry for responding late, was caught up in my university mid semester exams and they've finally ended. I've added it :)

> That makes sense. My question is if we need concern about the case where the user gives permission through app's settings, do we also not need to concern ourselves about the case where user clicks "Yes" and later revokes permission through the settings screen. In this case the `doNotAskForLocationPermission` should be set to `true` from NearbyParentFragment and ExploreMapFragment too right?

Um, figuring out whether it's actually the user who has revoked the permission from settings would be difficult, as Android turns off permission automatically for unused apps. So in this case if we set `doNotAskForLocationPermission` to true, then we will never be able to ask for permission even when user hasn't himself denied or turned it off.

Reply:
> 
> Um, figuring out whether it's actually the user who has revoked the permission from settings would be difficult, as Android turns off permission automatically for unused apps. So in this case if we set `doNotAskForLocationPermission` to true, then we will never be able to ask for permission even when user hasn't himself denied or turned it off.

I know it is not always possible to ensure the global state is set correctly, The ideal solution is to just do away with the global state. I opened #5678 about the same.

I was of belief that we could set the `doNotAskForLocationPermission` in some relevant place reliably in the new code flow. Since this PR has been long standing I'll rest my case, though.

Reply:
No problem at all, @sivaraam. I, too, would love to remove this global state by working on that issue as soon as this gets merged :)